pkg/chunked/storage_linux.go

```diff
@@ -1339,11 +1339,14 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 	}
 
 	filesToWaitFor := 0
-	for i, r := range mergedEntries {
+	for i := range mergedEntries {
+		r := &mergedEntries[i]
 		if options.ForceMask != nil {
 			value := idtools.FormatOverrideXattr(r.UID, r.GID, int(r.Mode))
-			mergedEntries[i].Xattrs[idtools.ContainersOverrideXattr] = base64.StdEncoding.EncodeToString([]byte(value))
-			r.Mode = int64(*options.ForceMask)
+			if r.Xattrs == nil {
+				r.Xattrs = make(map[string]string)
+			}
+			r.Xattrs[idtools.ContainersOverrideXattr] = base64.StdEncoding.EncodeToString([]byte(value))
```

Comment:
It feels a bit unclean to mutate the input here; I'd trust you if you said it was safe, but still.

In ostree we have something very similar to the `ContainersOverrideXattr` in the "bare-user" mode vs "bare" and the code paths in the backend are pretty distinct. Doing something similar here would want instead for e.g. `setFileAttrs` to take an override parameter or so perhaps.

Reply:
This should be safe because `mergedEntries` is an array of values (i.e. copies), created from scratch for this function. … ugh, except that a hash is a reference type, so AFAICS this _might_ be affecting the `c.toc` field.

Ultimately, that is _still_ safe because `c.toc` is written at most once, and read exactly once; but the number of assumptions _is_ getting a bit large.

Reply:
In Rust:

- xattrs would likely be an `Option<T>` instead of a pointer, so the compiler would have made us check for its presence
- We likely would have been passing a `&TableOfContents` or so, and the compiler would have errored out on an attempt to mutate it
- And if we chose to pass instead `&mut TableOfContents` the compiler would verify that there weren't other threads (goroutines) that were concurrently reading it, etc.
- But we can often do something better than mutation for something like this; Rust has efficient iterators over collections, and in this case instead of mutation it'd likely be better to leave the input read-only and instead pass `impl Iterator<Item=Xattr>` to the function that sets xattrs (instead of an array). It's efficient to do e.g. `toc.xattrs.iter().chain(Xattr::new(idtools.CONTAINER_OVERRIDE_XATTR, value))` (although this would blend borrowed and owned values so we may need a `Cow` or so)

Reply:
I’d sign up to be a platinum member of the Go haters club, but that’s not really under discussion in this PR. (If you wanted to make a case for rewriting the stack in Rust, I’d not oppose that at all.)

---

I do agree that _this_ would be safer with a deep copy, but then … Go doesn’t have a language-native deep copy operation _either_. So it’s either risk of bugs due to overwriting data (if it ever turned from single use into multiple-use) or risk of bugs due to missing a field copy (if we ever added a field), and to me that feels like a wash, up to @giuseppe .